Fix for Rancher Service Discovery and adding custom tag support

alextbrandwatch opened the pull request:
### What does this PR do?

- Restores Service Discovery functionality in Rancher in the event that Docker container metadata isn't populated as expected.
- Incorporates functionality previously created by @janeczku at Rancher to automatically tag hosts based on their labels in Rancher.

### Motivation

In our experience running a modified version of the Datadog agent in production, we've seen that Rancher is not guaranteed to populate Docker container metadata correctly as seen with the `docker inspect` command.  We believe this is a bug with Rancher which we have separately raised with them.  However, the result is that Service Discovery does not always work on the current release of Rancher, which is a major issue.

To address this, we build on the work recently contributed by @zippolyte and in our own fork to introduce an integration with the Rancher Metadata API.  When resolving host IP address and exposed ports for the purposes of Service Discovery, this PR adds the ability to read these from the Rancher Metadata API should all previous attempts to resolve these values fail.

In addition, we've taken this opportunity to integrate some of the work previously carried out by @janeczku with this customisation of the agent: https://github.com/janeczku/datadog-rancher-init.  We've added a new configuration file flag: `collect_rancher_host_labels`.  With this flag set to `true`, the agent will check for connectivity to the Metadata API of a local Rancher environment, and if successful will read the metadata labels of the host from the API and set each entry as a tag on the host in Datadog.  See the below example of tags automatically inferred in this way:

<img width="535" alt="screen shot 2017-06-01 at 21 10 29" src="https://cloud.githubusercontent.com/assets/17849720/26698714/f655282a-470e-11e7-9a56-1ea0359b63c4.png">

The net result is that Service Discovery is seen to be much more stable when running under Rancher, and Rancher users can deploy the "vanilla" agent directly from Datadog whilst retaining the functionality from [datadog-rancher-init](https://github.com/janeczku/datadog-rancher-init).

### Additional Notes

Before this PR can be merged, the following PRs must be merged into their respective repos.  These PRs constitute a single piece of work.

- dd-agent-omnibus: https://github.com/DataDog/dd-agent-omnibus/pull/176
(adding the `rancher-metadata` dependency, source: https://github.com/m4ce/rancher-metadata-python)
- omnibus-software: https://github.com/DataDog/omnibus-software/pull/147
(adding `rancher-metadata` dependency)

This PR has a co-dependency with this one:

- integrations-core: https://github.com/DataDog/integrations-core/pull/455
(updating references to variables renamed and relocated in this PR)

Finally, this PR is dependent upon this one:

- docker-dd-agent: https://github.com/DataDog/docker-dd-agent/pull/206
(adding environment variable `RANCHER_HOST_LABELS` to update the `datadog.conf` file)

hkaj commented on utils/service_discovery/sd_docker_backend.py:
```diff
@@ -317,15 +348,39 @@ def get_tags(self, state, c_id):
+                    tags.append('rancher_host_docker_version:%s' % v)
+
+                elif k == RancherUtil.HOST_LINUX_KERNEL_VERSION_LABEL:
+                    tags.append('rancher_host_linux_kernel_version:%s' % v)
```
I don't think those 3 fit as tags in service discovery. They're good as host metadata but could you remove them from here?

alextbrandwatch replied:
Yes I wondered about this - will remove.  Thanks!

alextbrandwatch commented on utils/rancher/rancherutil.py:
```diff
@@ -0,0 +1,69 @@
+    def is_rancher():
+        if RancherUtil._is_rancher is None:
+            try:
+                response = requests.get(url=RancherUtil.METADATA_URL)
```
1 second seems sensible, I'm fairly sure the request goes to another container on the same machine.  I've never known it not respond immediately.

alextbrandwatch commented:
Hi @hkaj, thank you for the speedy review.  I've made the changes as requested: relocating `rancherutil.py` into the `orchestrator` folder and removing the unnecessary container tags.  I've built and re-deployed the agent internally and everything looks good from what I can see.

I've also pushed up changes based on your comments on the other PRs.

Thanks again - I appreciate your help!

alextbrandwatch commented:
Thanks so much! 🎉

janeczku commented:
@alextbrandwatch so great! 👏  🎉

hkaj commented:
Hey @alextbrandwatch sorry we released a change that conflicts with your PR, I'll rebase your commits shortly.

alextbrandwatch commented:
Thanks for the heads up @hkaj, looks like a bit of a change to how orchestration environments are detected and managed, a rebase would be appreciated though I will take a look myself when I'm next able to 🙇

alextbrandwatch commented:
Hi @hkaj, I've rebased this PR against the changes that were made to the way that Orchestrator utility classes were handled.  The result is that this PR is now a little simpler.

I've also removed the `check_rancher_host_labels` configuration file option that I proposed, given that a `check_orchestrator_label` option has now been added.

I'll get this tested and deployed to our test environment as soon as possible, but in the meantime I would very much appreciate a re-review to see if I've missed anything.

Thank you!

hkaj commented on utils/orchestrator/rancherutil.py:
```diff
@@ -0,0 +1,102 @@
+        return tags
+
+    @staticmethod
+    def is_rancher():
```
any reason not to rename it to is_detected and get rid of L29-31?

hkaj commented on utils/orchestrator/rancherutil.py:
```diff
@@ -0,0 +1,102 @@
+    def is_detected():
+        return RancherUtil.is_rancher()
+
+    def _get_cacheable_tags(self, cid=None, co=None):
```
also maybe make the default value for `co` `{}`? Otherwise `co.get` will error out

masci commented:
Hi @alextbrandwatch,

thanks for your contribution. We've discussed this and while we agree the feature has merit,  we decided to dedicate engineering resources to the [new Agent](https://github.com/DataDog/datadog-agent) thus limiting the work on this release line to bugfixes only.
